### ISSUE TITLE
Bug: react/prop-types throws missing props on chain propTypes passing

### ISSUE THREAD

**Issue by lightbringer1991** [yannickcr/eslint-plugin-react#1439]
```jsx
import React from 'react';
import PropTypes from 'prop-types';

class XYZ extends React.Component {
  render() {
    this.props.onClick();
  }
}

class ABC extends React.Component {
  render() { return null; }
}

ABC.propTypes = {
  onClick: PropTypes.func.required,
};

XYZ.propTypes = ABC.propTypes;
```

apparently `eslint` complains that XYZ doesn't have `onClick` props

**Comment by ljharb** [comment]
There's no practical way for a linter to reliably follow that kind of dynamism around your file.

propTypes serve as documentation, and ABC and XYZ's propType objects should be distinct. Try this:

```js
XYZ.propTypes = {
  onClick: ABC.propTypes.onClick,
};
```

**Comment by sonhanguyen** [comment]
@ljharb I don't know anything about ast parsing but shouln't there be a way to just eval this `propTypes` expression and check the result? Expect it to come with some performance cost however it will accommodate reusing `propTypes` better and `propTypes` seems always statically resolvable so maybe it's not much cost?
Again total noobs, if you think it's worth a shot I'll be happy to learn how.

**Comment by ljharb** [comment]
No, "eval" isn't something any linter can do.

Separately, making new objects for propTypes once per component has no measureable performance cost, but not doing it can have measureable clarity cost.

**Comment by sonhanguyen** [comment]
https://github.com/yannickcr/eslint-plugin-react/issues/476
